docs: add direct IO section to architecture (#2770)
tigerbeetle/tigerbeetle: docs/internals/ARCHITECTURE.md

also plays a key role in the internal implementation. Every object in TigerBeetle has a globally
unique `u64` creation timestamp, which plays the role of synthetic primary key.

### Direct IO

> TigerBeetle bypasses operating system's page cache and uses Direct IO. That is, normally, when an

Reviewer: ```suggestion
TigerBeetle bypasses the operating system's page cache and uses Direct IO. That is, normally, when an
```

application writes to a file, the operating system only updates the in-memory cache, and the data
gets to disk later. This is a good default for the vast majority of the applications. TigerBeetle is
an exception --- it instructs the operating system to read directly from the disk, and write directly
to the disk, bypassing any caches (please refer to this
[excellent article](https://transactional.blog/how-to-learn/disk-io) for the overview of the
relevant OS APIs).

> Bypassing page cache is required for correctness. While operating systems provide `fsync` API to

Reviewer: ```suggestion
Bypassing page cache is required for correctness. While operating systems provide `fsync` APIs to
```

or

```suggestion
Bypassing page cache is required for correctness. While operating systems provide an `fsync` API to
```

> Bypassing page cache is required for correctness. While operating systems provide `fsync` API to

Reviewer: ```suggestion
Bypassing the page cache is required for correctness. While operating systems provide `fsync` API to
```

(but, this one reads OK as "bypassing page cache" too I think, just later we refer to "TigerBeetle doesn't require the OS to provide a page cache" which could also be "TigerBeetle doesn't require the OS to provide page cache")

Author: dropped both articles!

flush page cache to disk, it doesn't allow handling errors reliably:
[Can Applications Recover from fsync Failures?](https://www.usenix.org/system/files/atc20-rebello.pdf)

The second reason to bypass the cache is the general principle of avoiding dependencies and assuming
less about the interface. Concretely, TigerBeetle doesn't require the OS to provide a page cache,
and it doesn't require a file system by virtue of using only a single file. As a consequence,
TigerBeetle can run directly against a block device.

> ### Flexible Quorums

Reviewer: Great additions, these two sections!

There's something odd about a TigerBeetle cluster --- we recommend using an even number of replicas,
`6`. Usually, consensus implementations use `2f + 1` nodes, `3` or `5`, to have a clear majority for
quorums. TigerBeetle uses so-called flexible quorums. For `6` replicas, the replication quorum is

> only `3`. That is, it is enough for only a half of the cluster to persist a prepare durably to disk

Reviewer: I see you like your `That is` :smile:. It's a personal thing, but I think we could drop it to be more direct!

for it to be considered logically committed. On the other hand, for changing views (that is, rotating
the role of the primary to a different replica), at least `4` replicas are needed. Because any
subset of three replicas and any subset of four replicas intersect, the new primary is guaranteed to
know all potentially committed prepares.

The upshot here is that it's enough for `3` replicas to be online, including the primary, to keep
the cluster available. If in a cluster of `6`, three replicas crash simultaneously, there's a
50% chance that the cluster remains available. If the replicas are crashing one by one, the chance
is higher: when only `4` replicas remain, the chance that the next one to crash would be a primary
is only 25%.

### Protocol Aware Recovery

TigerBeetle assumes that the disk can fail, and that the data can be physically lost eventually even

> if the original `write` and `fsync` completed successfully. If that happens, TigerBeetle

Reviewer: We mention how we don't rely on `fsync` above, so maybe something like

```suggestion
if the original `write` (with `O_DSYNC`) completed successfully. If that happens, TigerBeetle
```

Author: I think I'd rather leave this one as is: I am speaking abstractly about the disk/OS API here, rather than our specific way of using them.
